File: apps/web/src/lib/garden.ts
// Reading-garden gamification — data layer.
//
// Wired to the backend at /gardens/*. The shapes here match the camelCased
// wire format produced by app/schemas/garden.py.

import { api } from "./api";

export type SpeciesId = "ficus" | "helianthus" | "lavandula" | "monstera";

export interface Species {
  id: SpeciesId;
  name: string;
  latin: string;
  blurb: string;
  unlocked: boolean;
}

export const SPECIES: Species[] = [
  { id: "ficus",      name: "Ficus litteraria", latin: "novel-fig",       blurb: "Rewards long, immersive sessions.", unlocked: true  },
  { id: "helianthus", name: "Helianthus",       latin: "poet's sunflower", blurb: "Loves daily streaks.",              unlocked: true  },
  { id: "lavandula",  name: "Lavandula",        latin: "essay-lavender",   blurb: "Thrives on careful note-taking.",   unlocked: true  },
  { id: "monstera",   name: "Monstera",         latin: "epic-leaf",        blurb: "Unlocks after your first finished book.", unlocked: false },
];

export type FarmerHat   = "straw" | "wool" | "scholar" | "none";
export type FarmerCoat  = "denim" | "linen" | "earth"  | "moss";
export type FarmerSkin  = "fair" | "tan" | "umber" | "sepia";
export type FarmerTool  = "watering-can" | "shears" | "lantern" | "book";

export interface Farmer {
  hat: FarmerHat;
  coat: FarmerCoat;
  skin: FarmerSkin;
  tool: FarmerTool;
  name: string;
}

/** 0 (seed) → 6 (full bloom). Stage IV = bud forming. */
export type GrowthStage = 0 | 1 | 2 | 3 | 4 | 5 | 6;

export type GardenHealth = "thriving" | "budding" | "wilting" | "dying";

export type EventKind = "read" | "quiz" | "water" | "skip" | "translate" | "tend";

export interface JournalEntry {
  id: string;
  at: string;          // ISO
  kind: EventKind;
  what: string;        // already formatted by the server; html subset (`<em>`)
  delta: string;       // e.g. "+ 6 leaves", "− 1 droplet"
  warn?: boolean;
}

export interface Garden {
  bookId: string;
  bookTitle: string;
  bookAuthor: string | null;
  startedAt: string;       // ISO
  species: SpeciesId;
  farmer: Farmer;

  stage: GrowthStage;
  growthPercent: number;
  pagesRead: number;
  pageCount: number;
  pagesReadDelta: number;
  quizzesAnswered: number;
  quizzesTotal: number;
  quizAccuracyPercent: number;

  vitality: number;
  vitalityCapacity: number;
  daysUntilThirst: number;
  weeklyTendingDueAt: string;

  streakDays: number;
  bestStreakDays: number;
  newLeaves: number;
  lastLeafAt: string | null;

  journal: JournalEntry[];
}

export interface GardenSummary {
  bookId: string;
  bookTitle: string;
  bookAuthor: string | null;
  species: SpeciesId;
  stage: GrowthStage;
  growthPercent: number;
  health: GardenHealth;
}

// --- Weekly tending -------------------------------------------------------

export interface TendingQuestion {
  id: string;
  prompt: string;
  choices: string[];
  /** Server strips these pre-submit; only present in mock fixtures. */
  correctIndex?: number;
  explanation?: string;
}

export interface TendingResult {
  score: number;
  total: number;
  passed: boolean;
  growthGained: number;
  vitalityRestored: number;
  newStage: GrowthStage;
  nextDueAt: string;
  perQuestion: {
    id: string;
    correct: boolean;
    givenIndex: number;
    /** Revealed only in the post-submit response. */
    correctIndex?: number;
    explanation?: string;
  }[];
}

// ---------------------------------------------------------------------------
// API
// ---------------------------------------------------------------------------

export async function getGarden(bookId: string): Promise<Garden> {
  return api<Garden>(`/gardens/${bookId}`);
}

export async function listGardens(): Promise<GardenSummary[]> {
  return api<GardenSummary[]>(`/gardens`);
}

export async function updateGarden(
  bookId: string,
  body: { species?: SpeciesId; farmer?: Farmer },
): Promise<Garden> {
  return api<Garden>(`/gardens/${bookId}`, {
    method: "PATCH",
    body,
  });
}

export async function recordEvent(
  bookId: string,
  kind: EventKind,
  payload: Record<string, unknown> = {},
): Promise<Garden> {
  return api<Garden>(`/gardens/${bookId}/events`, {
    method: "POST",
    body: { kind, payload },
  });
}

export async function getTendingQuestions(bookId: string): Promise<TendingQuestion[]> {
  return api<TendingQuestion[]>(`/gardens/${bookId}/tending`);
}

export async function submitTending(
  bookId: string,
  answers: { questionId: string; choiceIndex: number }[],
): Promise<TendingResult> {
  return api<TendingResult>(`/gardens/${bookId}/tending`, {
    method: "POST",
    body: { answers },
  });
}

// ---------------------------------------------------------------------------
// Display helpers
// ---------------------------------------------------------------------------

/** "Day 14 of cultivation" */
export function daysSince(iso: string): number {
  const ms = Date.now() - new Date(iso).getTime();
  return Math.max(1, Math.floor(ms / 86_400_000));
}

/** Format an ISO date as "11 May · 09:14" */
export function formatJournalDate(iso: string): string {
  const d = new Date(iso);
  const day = d.getDate().toString().padStart(2, "0");
  const month = d.toLocaleString("en", { month: "short" });
  const time = d.toTimeString().slice(0, 5);
  return `${day} ${month} · ${time}`;
}

/** Countdown formatted as HH:MM:SS, or "00:00:00" if past. */
export function countdown(iso: string, nowMs = Date.now()): string {
  const ms = new Date(iso).getTime() - nowMs;
  if (ms <= 0) return "00:00:00";
  const totalSeconds = Math.floor(ms / 1000);
  const hh = Math.floor(totalSeconds / 3600).toString().padStart(2, "0");
  const mm = Math.floor((totalSeconds % 3600) / 60).toString().padStart(2, "0");
  const ss = (totalSeconds % 60).toString().padStart(2, "0");
  return `${hh}:${mm}:${ss}`;
}

File: apps/api/app/schemas/garden.py
"""Pydantic schemas for garden endpoints.

Keep field names in sync with apps/web/src/lib/garden.ts — the frontend uses
camelCase via populate_by_name aliasing so the wire format stays JS-friendly.
"""
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

GardenSpeciesLiteral = Literal["ficus", "helianthus", "lavandula", "monstera"]
GardenHealthLiteral = Literal["thriving", "budding", "wilting", "dying"]
GardenEventKindLiteral = Literal["read", "quiz", "water", "skip", "translate", "tend"]
FarmerHat = Literal["straw", "wool", "scholar", "none"]
FarmerCoat = Literal["denim", "linen", "earth", "moss"]
FarmerSkin = Literal["fair", "tan", "umber", "sepia"]
FarmerTool = Literal["watering-can", "shears", "lantern", "book"]


def _alias(name: str) -> str:
    """snake_case → camelCase for the JSON wire format."""
    parts = name.split("_")
    return parts[0] + "".join(p.title() for p in parts[1:])


class _CamelModel(BaseModel):
    model_config = ConfigDict(alias_generator=_alias, populate_by_name=True, from_attributes=True)


class FarmerSchema(_CamelModel):
    hat: FarmerHat = "straw"
    coat: FarmerCoat = "earth"
    skin: FarmerSkin = "tan"
    tool: FarmerTool = "watering-can"
    name: str = ""


class JournalEntry(_CamelModel):
    id: uuid.UUID
    at: datetime
    kind: GardenEventKindLiteral
    what: str
    delta: str
    warn: bool = False


class GardenSummary(_CamelModel):
    book_id: uuid.UUID
    book_title: str
    book_author: str | None = None
    species: GardenSpeciesLiteral
    stage: int
    growth_percent: int
    health: GardenHealthLiteral


class GardenRead(_CamelModel):
    book_id: uuid.UUID
    book_title: str
    book_author: str | None = None
    started_at: datetime
    species: GardenSpeciesLiteral
    farmer: FarmerSchema

    stage: int
    growth_percent: int
    pages_read: int
    page_count: int = 0
    pages_read_delta: int = 0
    quizzes_answered: int
    quizzes_total: int = 0
    quiz_accuracy_percent: int = 0

    vitality: int
    vitality_capacity: int
    days_until_thirst: int
    weekly_tending_due_at: datetime

    streak_days: int
    best_streak_days: int
    new_leaves: int
    last_leaf_at: datetime | None = None

    journal: list[JournalEntry] = Field(default_factory=list)


class GardenUpdate(_CamelModel):
    species: GardenSpeciesLiteral | None = None
    farmer: FarmerSchema | None = None


# --- events --------------------------------------------------------------

class RecordEventRequest(_CamelModel):
    kind: GardenEventKindLiteral
    payload: dict = Field(default_factory=dict)


# --- weekly tending ------------------------------------------------------

class TendingQuestion(_CamelModel):
    id: str
    prompt: str
    choices: list[str]
    # `correctIndex` and `explanation` are present in the mock; in the real
    # backend the server holds them privately and these are absent on the
    # wire pre-submit. Kept optional for forward-compat with the mock client.
    correct_index: int | None = None
    explanation: str | None = None


class SubmitTendingAnswer(_CamelModel):
    question_id: str
    choice_index: int


class SubmitTendingRequest(_CamelModel):
    answers: list[SubmitTendingAnswer]


class TendingPerQuestion(_CamelModel):
    id: str
    correct: bool
    given_index: int
    # Revealed only post-submit so the review screen can show the right
    # answer and the explanation. Absent in the pre-submit GET payload.
    correct_index: int | None = None
    explanation: str | None = None


class TendingResult(_CamelModel):
    score: int
    total: int
    passed: bool
    growth_gained: int
    vitality_restored: int
    new_stage: int
    next_due_at: datetime
    per_question: list[TendingPerQuestion]
